numpy/ma/core.py

```diff
@@ -5079,24 +5007,66 @@ def mean(self, axis=None, dtype=None, out=None):
         1.5
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = self._data.mean(axis, dtype, out, **kwargs)
+
+            to_mask = ~np.isfinite(ret)
+            if self.size == 0:
+                to_mask = True
+
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    out.__setmask__(to_mask if np.any(to_mask) else False)
+                return out
+
+            if isinstance(ret, ndarray):
+                ret = ret.view(type(self))
+                ret.__setmask__(to_mask if np.any(to_mask) else False)
+                return ret
+            else:
+                # otherwise it's a scalar
+                return ret if self.size else masked
+
+        # code largely copied from _mean in numpy/core/_methods.py
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
```

Comment:
Ah I see, it's a bit trickier because `out` might be a plain ndarray, which would cause `ret` to be one too, which causes test failures. But with an extra change below I think I got it to work.